### PR TITLE
Add constraint and config options

### PR DESCRIPTION
**Pull request by AlexAvlonitis**
- Adds config for warden to not intercept 401 requests
  This config allows apps to change the default behaviour of intercepting 401 requests, by including an initialiser:
  `GDS::SSO.config { |config| config.intercept_401_responses = false }`.
  For example apps that also include basic http auth would probably need this option.
  
 - Adds authorised user constraint so that consumer apps can easily add a permission based constraint.
 
 - Add PermissionDeniedError error class under GDS::SSO namespace and a warning log message that the existing PermissionDeniedException should be deprecated on a new major release.
 
 - Extract the user authorisation logic into a separate class and Re-use authorisation logic in the controller_methods class
 
 more info : https://trello.com/c/qb1qpwnr/1369-additional-functionality-to-gds-sso-gem-for-intercepting-401s-and-providing-a-route-constraint

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,5 +4,5 @@ gemfiles/*.lock
 Gemfile.lock
 pkg/*
 tmp/*
-*.sqlite
+*.sqlite*
```
Comment (kevindew) [inline .gitignore:7]: This would be good to be a separate PR is it related to: https://github.com/alphagov/gds-sso/pull/299 ?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,12 @@
 ## Unreleased
 
+# 19.1.0
+
+* Adds config option that allows apps to change the default behaviour of intercepting 401 requests, by including an initialiser: `GDS::SSO.config { |config| config.intercept_401_responses = false }`.
+* Added AuthorisedUserConstraint class so that consumer apps can easily add a permission based constraints.
+* Adds depracation warning for PermissionDeniedException to be replaced by PermissionDeniedError
```
Comment (kevindew) [inline CHANGELOG.md:7]:
```suggestion
* Adds deprecation warning for GDS::SSO::ControllerMethods::PermissionDeniedException. If your app uses this you can replace it with GDS::SSO::PermissionDeniedError.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,12 @@
...
+* Extracted User authorisation logic residing in ControllerMethods class into a separate class `AuthoriseUser` so it can be reused by both the ControllerMethods and the AuthorisedUserConstraint
```
Comment (kevindew) [inline CHANGELOG.md:8]:
```suggestion
```

Since this is an implementation detail and not a class we expect consumers to use we don't need to put it in the changelog.

**Review by kevindew — COMMENTED** [review]
Thanks Alex, this looks great. I've not had a chance to look at tests yet but thought I'd hit submit.

File: spec/controller/controller_methods_spec.rb [hunk 4]
```diff
@@ -1,57 +1,24 @@
...
-
-      expect { ControllerSpy.new(current_user).authorise_user!(any_of: %w[good bad]) }.to raise_error(expected_error)
-    end
-  end
-
-  context "with none of `any_of` or `all_of`" do
-    it "raises an `ArgumentError`" do
-      expect { ControllerSpy.new(current_user).authorise_user!(whoops: "bad") }.to raise_error(ArgumentError)
+        expect { ControllerSpy.new(current_user).authorise_user!("bad") }.to raise_error(expected_error)
+      end
```
Comment (kevindew) [inline spec/controller/controller_methods_spec.rb:21]: Looks good 👍

File: .gitignore (hunk 1 above)
Comment (AlexAvlonitis) [reply .gitignore:7]: Not sure if it's related, it's more related to the combustion gem, when you run rspec, combustion creates couple of sqlite instances and appends random chars at the end `...sqlite-asd`, so just adding it in the gitignore not to accidentally commit them

File: lib/gds-sso/config.rb [hunk 5]
```diff
@@ -29,6 +29,9 @@ module Config
 
       mattr_accessor :api_only
 
+      mattr_accessor :intercept_401_responses
+      @@intercept_401_responses = true
```
Comment (kevindew) [inline lib/gds-sso/config.rb:33]: It'd be good to have this tested could we add something to the config spec?

File: lib/gds-sso/authorised_user_constraint.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+module GDS
+  module SSO
+    class AuthorisedUserConstraint
```
Comment (kevindew) [inline lib/gds-sso/authorised_user_constraint.rb:3]: It'd be good to have an integration test for this.

I'd suggest adding this around a route in: https://github.com/alphagov/gds-sso/blob/main/spec/internal/config/routes.rb and adding a system test that checks a non logged in request, a request without necessary permission and one with necessary permission

**Review by kevindew — COMMENTED** [review]
Just added some testing notes. The tests look good but I don't think we have yet got sufficient testing.

File: lib/gds-sso/authorised_user_constraint.rb (hunk 6 above)
Comment (AlexAvlonitis) [reply lib/gds-sso/authorised_user_constraint.rb:3]: I'm trying this: [0aa3a1a](https://github.com/alphagov/gds-sso/pull/300/commits/0aa3a1aa5f5fb0ce405b8385c0c2247299892d73) but it is giving me hard time. Just posted the WIP changes in case you can see what i'm doing wrong while i'm debugging.

Seems like the constraint errors are not rescued by the controller_methods or the railtie, the system spec is raising the error without rescuing. The combustion gem is supposed to include the railtie as well, i tried to run the railtie manually too but no luck

Comment (kevindew) [reply lib/gds-sso/authorised_user_constraint.rb:3]: So Rails needs to be configured to handle the resuce responses as I understand.

If I change [`Combustion.initialize!`](https://github.com/alphagov/gds-sso/blob/84da55d0ce05197c695af776c8b26b04d68af09b/spec/spec_helper.rb#L9-L11) to be:

```
Combustion.initialize! :all do
  config.cache_store = :null_store
  config.action_dispatch.show_exceptions = true
end
```

I can get HTTP responses. However I'm getting a 500 rather than a 403 so just looking into why

Comment (AlexAvlonitis) [reply lib/gds-sso/authorised_user_constraint.rb:3]: Yes this option I think renders the actual error page instead of stopping on raise. So I think you get the classic red  rails error page, saying: `GDS::SSO::PermissionDeniedError in ExampleController#constraint_restricted\nSorry, you don't seem to have the execute permission for this app.\n`. with 500 status code.

Comment (kevindew) [reply lib/gds-sso/authorised_user_constraint.rb:3]: I did also find the test works if we change Combustion::Initialize! to:

```
Combustion.initialize! :all do
  config.cache_store = :null_store
  config.action_dispatch.show_exceptions = true
  config.action_dispatch.rescue_responses.merge!(
    "GDS::SSO::PermissionDeniedError" => :forbidden
  )
end
```

so not sure our code is initialising in the right place. Does the 403 work in govuk-chat? i.e. is this a quirk of the test suite or a feature that isn't working.

Comment (kevindew) [reply lib/gds-sso/authorised_user_constraint.rb:3]: I also notice it works if we move the code out the initializer block: 

```
    class Railtie < Rails::Railtie
      initializer "gds-sso.initializer" do
        GDS::SSO.config do |config|
          config.cache = Rails.cache
          config.api_only = Rails.configuration.api_only
        end
        OmniAuth.config.logger = Rails.logger
      end

      config.action_dispatch.rescue_responses.merge!(
        "GDS::SSO::PermissionDeniedError" => :forbidden,
      )
    end
```

Though that still seems a bit wrong. Will do some checking in rails code

Comment (kevindew) [reply lib/gds-sso/authorised_user_constraint.rb:3]: Looking at the ActiveRecord Railtie it is configured as above:  https://github.com/rails/rails/blob/1d13a7c44d205a67cdd692bad56399e41a4eea44/activerecord/lib/active_record/railtie.rb#L23-L28

So I think I just gave you lousy advice in: https://github.com/alphagov/gds-sso/pull/300#discussion_r1575876567

Comment (AlexAvlonitis) [reply lib/gds-sso/authorised_user_constraint.rb:3]: Thanks Kevin 🚀

File: lib/gds-sso/controller_methods.rb [hunk 7]
```diff
@@ -1,11 +1,19 @@
 module GDS
   module SSO
+    class PermissionDeniedError < StandardError
+    end
+
     module ControllerMethods
-      class PermissionDeniedException < StandardError
+      # TODO: remove this for the next major release
+      class PermissionDeniedException < GDS::SSO::PermissionDeniedError
```
Comment (kevindew) [inline lib/gds-sso/controller_methods.rb:8]:
```suggestion
      class PermissionDeniedException < PermissionDeniedError
```

Just realised since the name of the exception is not a clash we don't need the `GDS::SSO` prefix

File: lib/gds-sso/controller_methods.rb [hunk 8]
```diff
@@ -1,11 +1,19 @@
...
+        def initialize(...)
+          warn "GDS::SSO::ControllerMethods::PermissionDeniedException is deprecated, please replace with GDS::SSO::PermissionDeniedError"
+          super(...)
+        end
       end
 
       def self.included(base)
-        base.rescue_from PermissionDeniedException do |e|
+        base.rescue_from GDS::SSO::PermissionDeniedError do |e|
```
Comment (kevindew) [inline lib/gds-sso/controller_methods.rb:16]:
```suggestion
        base.rescue_from PermissionDeniedError do |e|
```

File: lib/gds-sso/railtie.rb [hunk 9]
```diff
@@ -1,6 +1,10 @@
 module GDS
   module SSO
     class Railtie < Rails::Railtie
+      config.action_dispatch.rescue_responses.merge!(
+        "GDS::SSO::PermissionDeniedError" => 403,
```
Comment (kevindew) [inline lib/gds-sso/railtie.rb:5]:
```suggestion
        "GDS::SSO::PermissionDeniedError" => :forbidden,
```

To match how Rails configures these

File: spec/spec_helper.rb [hunk 10]
```diff
@@ -8,6 +8,7 @@
 
 Combustion.initialize! :all do
   config.cache_store = :null_store
+  config.action_dispatch.show_exceptions = :all
```
Comment (kevindew) [inline spec/spec_helper.rb:11]:
```suggestion
  config.action_dispatch.show_exceptions = true
```

Do we need :all here? I imagine a boolean would be easier for someone to follow. It'd also be good to have either the commit message or a comment document why we've had to turn this on

File: spec/unit/config_spec.rb [hunk 11]
```diff
@@ -25,4 +25,10 @@
       end
     end
   end
+
+  describe ".intercept_401_responses" do
+    it "defaults to true" do
+      expect(subject.intercept_401_responses).to eq(true)
+    end
```
Comment (kevindew) [inline spec/unit/config_spec.rb:32]: Hmm it feels like just this isn't really testing this config option sufficiently. Perhaps we need a test in the railtie file that asserts that this config option actually does something. What do you think?

**Review by kevindew — COMMENTED** [review]
Great stuff Alex, this looks really close. Once we've got my comments resolved we'll seek final review from the owning team.

The only thing I notice is missing is the CHANGELOG entry - we can do that before we release the gem, it'll just go under the "Unreleased" header.

File: spec/spec_helper.rb (hunk 10 above)
Comment (AlexAvlonitis) [reply spec/spec_helper.rb:11]: the true has been deprecated and it now recommends `:all`

File: lib/gds-sso/railtie.rb (hunk 9 above)
Comment (AlexAvlonitis) [reply lib/gds-sso/railtie.rb:5]: oops i changed that when i was testing

File: spec/spec_helper.rb (hunk 10 above)
Comment (kevindew) [reply spec/spec_helper.rb:11]: Ah super.

File: lib/gds-sso/controller_methods.rb (hunk 7 above)
Comment (AlexAvlonitis) [reply lib/gds-sso/controller_methods.rb:8]: It won't have issues since this line is under GDS::SSO::ControllerMethods and the PermissionDeniedError is under GDS:SSO? The tests still pass when i change it, but i usually play it safe when we call things under different namespaces.

Comment (kevindew) [reply lib/gds-sso/controller_methods.rb:8]: Yup, we're already in the GDS::SSO namespace so need to qualify.

I suggested the namespace as I thought we'd have both GDS::SSO::PermissionDeniedException and GDS::SSO::ControllerMethods::PermissionDeniedException

File: spec/unit/config_spec.rb (hunk 11 above)
Comment (AlexAvlonitis) [reply spec/unit/config_spec.rb:32]: hmm yup good point, are you thinking of a test that we can set this value in a similar manner to the example of the description: `GDS::SSO.config { |config| config.intercept_401_responses = false }` ?

Comment (kevindew) [reply spec/unit/config_spec.rb:32]: More just something that asserts that this config value results in warden having intercept_401 switched on or not.

The sort of thing I'd want the test to verify is that if warden ever dropped the feature or renamed it this test would catch it.

Comment (AlexAvlonitis) [reply spec/unit/config_spec.rb:32]: i've added a testing approach, couldn't find a more elegant way, otherwise i would need to make an integration test.

Comment (kevindew) [reply spec/unit/config_spec.rb:32]: Came up with this, what do you think?

```
  describe "configuring intercept_401_responses" do
    it "sets warden intercept_401 to false when the configuration option is set to false" do
      GDS::SSO::Config.intercept_401_responses = false

      described_class.initializers.each(&:run)

      expect(warden_manager.config[:intercept_401]).to be(false)
    end

    it "sets warden intercept_401 to true when the configuration option is set to true" do
      GDS::SSO::Config.intercept_401_responses = true

      described_class.initializers.each(&:run)

      expect(warden_manager.config[:intercept_401]).to be(true)
    end
  end

  def warden_manager
    middleware = Rails.application.config.middleware.find { |m| m.name.include?("Warden::Manager") }

    Warden::Manager.new(nil, &middleware.block)
  end

```

Comment (kevindew) [reply spec/unit/config_spec.rb:32]: I set it up in a loop to run 1000 times and it didn't flake at all so I think it's ok 🤞

Comment (AlexAvlonitis) [reply spec/unit/config_spec.rb:32]: that's strange for me it does it like every 2-3 runs, 

image

Comment (kevindew) [reply spec/unit/config_spec.rb:32]: Though I see your code is perhaps different given the use of subject 😕

Comment (AlexAvlonitis) [reply spec/unit/config_spec.rb:32]: it's the `spec/unit/config_spec.rb` that failing, line 29. I suspect it's the random seed, it depends what test runs last it affects what the outcome is for that spec

Comment (kevindew) [reply spec/unit/config_spec.rb:32]: Ahh - sorry. I'd misunderstood that. 

Yes it makes sense that this one is flaky given that just has global vars. I don't really think the config test is sufficiently valuable to keep.

Though it is annoying to have tests that affect global state, so lets change the tests I suggested to mocks. e.g: 

```
        allow(GDS::SSO::Config).to receive(:intercept_401_responses).and_return(false)
```

**Comment by kevindew** [comment]
Great stuff. This all looks good to me. Let's see if the #govuk-publishing-platform team are happy. I'll drop them a ping.

File: lib/gds-sso/controller_methods.rb [hunk 12]
```diff
@@ -1,11 +1,19 @@
 module GDS
   module SSO
+    class PermissionDeniedError < StandardError
+    end
+
     module ControllerMethods
-      class PermissionDeniedException < StandardError
+      # TODO: remove this for the next major release
```
Comment (JonathanHallam) [inline lib/gds-sso/controller_methods.rb:7]: Could you maybe add a note to the commit message as to why this needs to be removed with the next major release?

Comment (AlexAvlonitis) [reply lib/gds-sso/controller_methods.rb:7]: will do, thanks for the review :)
